kyuubi-common/src/main/scala/org/apache/kyuubi/engine/deploy/EngineYarnModeSubmitter.scala

```diff
@@ -0,0 +1,451 @@
+
+    amContainer
+  }
+
+  private def prepareLocalResources(
+      destDir: Path,
+      env: mutable.HashMap[String, String]): mutable.HashMap[String, LocalResource] = {
+    info("Preparing resources for engine AM container")
+    // Upload kyuubi engine and the extra JAR to the remote file system if necessary,
+    // and add them as local resources to the application master.
+    val fs = destDir.getFileSystem(hadoopConf)
+
+    val localResources = mutable.HashMap[String, LocalResource]()
+    FileSystem.mkdirs(fs, destDir, new FsPermission(STAGING_DIR_PERMISSION))
+
+    distributeJars(localResources, env)
+    distributeConf(localResources)
+    localResources
+  }
+
+  private def distributeJars(
+      localResources: mutable.HashMap[String, LocalResource],
+      env: mutable.HashMap[String, String]): Unit = {
+    val jarsArchive = File.createTempFile(LOCALIZED_LIB_DIR, ".zip", Utils.createTempDir().toFile)
+    val jarsStream = new ZipOutputStream(new FileOutputStream(jarsArchive))
+    try {
+      jarsStream.setLevel(0)
+      val jars = kyuubiConf.getOption(KYUUBI_ENGINE_DEPLOY_YARN_MODE_JARS_KEY)
+      assert(jars.isDefined, "No jars specified for engine AM")
+      val putedEntry = new ListBuffer[String]
+      jars.get.split(KYUUBI_ENGINE_DEPLOY_YARN_MODE_ARCHIVE_SEPARATOR).foreach { path =>
+        val jars = Utils.listFilesRecursive(new File(path))
+        jars.foreach { f =>
+          if (!putedEntry.contains(f.getName) && f.isFile && f.getName.toLowerCase(
+              Locale.ROOT).endsWith(".jar") && f.canRead) {
+            jarsStream.putNextEntry(new ZipEntry(f.getName))
+            Files.copy(f.toPath, jarsStream)
+            jarsStream.closeEntry()
+            putedEntry += f.getName
+            addClasspathEntry(buildPath(Environment.PWD.$$(), LOCALIZED_LIB_DIR, f.getName), env)
+          }
+        }
+      }
+      putedEntry.clear()
+    } finally {
+      jarsStream.close()
+    }
+
+    distribute(
+      new Path(jarsArchive.getAbsolutePath),
+      resType = LocalResourceType.ARCHIVE,
+      destName = LOCALIZED_LIB_DIR,
+      localResources)
+  }
+
+  private def distributeConf(localResources: mutable.HashMap[String, LocalResource]): Unit = {
+    val confArchive = File.createTempFile(LOCALIZED_CONF_DIR, ".zip", Utils.createTempDir().toFile)
+    val confStream = new ZipOutputStream(new FileOutputStream(confArchive))
+    try {
+      confStream.setLevel(0)
+      val confs = kyuubiConf.getOption(KYUUBI_ENGINE_DEPLOY_YARN_MODE_CONF_KEY)
+      assert(confs.isDefined, "No conf specified for engine AM")
+      listDistinctFiles(confs.get).foreach { f =>
+        if (f.isFile && f.canRead) {
+          confStream.putNextEntry(new ZipEntry(f.getName))
+          Files.copy(f.toPath, confStream)
+          confStream.closeEntry()
+        }
+      }
+
+      val properties = confToProperties(kyuubiConf)
+      writePropertiesToArchive(properties, KYUUBI_CONF_FILE, confStream)
```

Comment:
there may be some disadvantages if we save and distribute engine conf to `kyuubi-defaults.conf`? for example, it's hard for user to understand the relationship between engine conf and server conf. I checked the Spark implementation, it also uses a different name `__spark_conf__.properties`